### PR TITLE
fix (module04-ex08) add of lambda_ in signature constructor + Example…

### PR DESCRIPTION
**Pull request by madvid**
Issue #167 
What have be done:
* Add of lambda_ in signature constructor of the class,
* a bit of code about it in the scope constructor
* Examples section to specify what is the expected behavior

Why:
There were 2 possible solutions (see issue #167), based on the instruction of the exercise 09 of the same module, first solution has been chosen.

### CODE REVIEW

**Comment by ezalos** [comment]
Your `if` `else` is unclear for me. 

Is this what you have in mind ?

```python
    if penality == "l2":
        self.lambda_ = lambda_
    else:
        self.lambda_ = 0
```
Such that other calculs simplify later on ?

```python
	def gradient(self, x, y):
		m = x.shape[0]

		hypothesis = self.predict_(x)
		loss = hypothesis - y
		grad = (x.T @ loss)

    
		theta_ = self.theta.copy()
		theta_[0] = 0
		regularization = self.lambda_ * theta_
		reg_gradient = (1/m) * (grad + regularization)

		return reg_gradient
```

It took me some time to figure it out. I'm worried we could have some students scratching their head on this question for longer than necessary.

If I correctly understood what you had in mind, I then advise for changing the example and be more specific by using the following block of code. I don't think we are revealing information important for them to figure out.

```python
class MyLogisticRegression():
	"""
	Description:
		My personnal logistic regression to classify things.
	"""
	def __init__(self, theta, alpha=0.001, max_iter=1000, penalty='l2', lambda_=1.0):
		# Check on type, data type, value ... if necessary
		self.alpha = alpha
		self.max_iter = max_iter
		self.theta = theta
		self.penalty = penalty
		if penality == "l2":
        	    self.lambda_ = lambda_
		else:
        	    self.lambda_ = 0
		... Your code ...

	... other methods ...
```

Also, another way student could come up with to accomplish the same result would be by having 2 different functions for the `loss()` and the `gradient()` and calling one of them depending on the value of `penalty`. 

As in this example, which might be faster:

```python
	def gradient(self, x, y):
		if self.penalty != 'l2':
			return self.gradient_no_reg(x, y)
		else:
			return self.gradient_reg(x, y)


	def gradient_reg(self, x, y):
		m = x.shape[0]

		hypothesis = self.predict_(x)
		loss = hypothesis - y
		grad = (x.T @ loss)

    
		theta_ = self.theta.copy()
		theta_[0] = 0
		regularization = self.lambda_ * theta_
		reg_gradient = (1/m) * (grad + regularization)

		return reg_gradient
	
	def gradient_no_reg(self, x, y):
		m = x.shape[0]

		hypothesis = self.predict_(x)
		loss = hypothesis - y
		grad = (x.T @ loss)

		_gradient = (1/m) * grad

		return _gradient
```

Our more precise `__init__()` might throw them off from exploring this idea.

**Comment by qfeuilla** [comment]
perso je trouve ca plus clean :
```python
class MyLogisticRegression():
	"""
	Description:
		My personnal logistic regression to classify things.
	"""
	def __init__(self, theta, alpha=0.001, max_iter=1000, penalty='l2', lambda_=1.0):
		# Check on type, data type, value ... if necessary
		self.alpha = alpha
		self.max_iter = max_iter
		self.theta = theta
		self.penalty = penalty
                supported_penalty = ["l2", ...]
                self.lambda_ = lambda_ if penalty in supported_penalty else 0
		... Your code ...

	... other methods ...
```

**Comment by qfeuilla** [comment]
un gros bloc de else if just pour ca c'est pas ouf, vaut mieux utiliser un ternaire

**Comment by ppeigne** [comment]
@qfeuilla : there is no other penalty than 'L2" in the bootcamp so it may be more confusing to present it as a list of options ? We could add a gentle comment in the code ?
 
```
class MyLogisticRegression():
	"""
	Description:
		My personnal logistic regression to classify things.
	"""
	def __init__(self, theta, alpha=0.001, max_iter=1000, penalty='l2', lambda_=1.0):
		# Check on type, data type, value ... if necessary
		self.alpha = alpha
		self.max_iter = max_iter
		self.theta = theta
		self.penalty = penalty
                supported_penalty = ["l2", ...] # We will consider 'l2' only but feel free to implement others penalty ;) 
                self.lambda_ = lambda_ if penalty in supported_penalty else 0
		... Your code ...

	... other methods ...
```
